clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

```diff
@@ -574,6 +583,10 @@ class CIRCastOpLowering : public mlir::OpConversionPattern<mlir::cir::CastOp> {
   mlir::LogicalResult
   matchAndRewrite(mlir::cir::CastOp castOp, OpAdaptor adaptor,
                   mlir::ConversionPatternRewriter &rewriter) const override {
+    // For arithmetic conversions, LLVM IR uses the same instruction to convert
+    // both individual scalars and entire vectors. This lowering pass handles
+    // both situations.
```

Comment:
Should `CastOp` be marked `ElementwiseMappable`?

Reply:
I don't know if `CastOp` should be `ElementwiseMappable`.  I am not familiar with `ElementwiseMappable`, and I am struggling to find documentation about its requirements and effects.  I found this in `mlir/docs/Traits/_index.md`:

> Note: Not all ops that are "elementwise" in some abstract sense satisfy this trait. In particular, broadcasting behavior is not allowed. See the comments on `OpTrait::ElementwiseMappable` for the precise requirements.

but I can't find `OpTrait::ElementwiseMappable` to know what the precise requirements are.

If `CastOp` should be `ElementwiseMappable`, then I assume that `BinOp` and some other operations should also be.

Reply:
> but I can't find `OpTrait::ElementwiseMappable` to know what the precise requirements are.

The two interesting uses I found are in `mlir/include/mlir/Dialect/Math/IR/MathOps.td` and in `mlir/include/mlir/Dialect/Vector/IR/VectorOps.td`. Also used in transformations in via `hasElementwiseMappableTraits`. I don't understand the full requirements either, even if we don't add it now we should probably keep an eye there.

> If `CastOp` should be `ElementwiseMappable`, then I assume that `BinOp` and some other operations should also be.

Yep.